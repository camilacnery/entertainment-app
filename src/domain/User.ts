export interface IUserProfile {
  imageUrl: string;
  fullName: string;
  firstName: string;
  email: string;
}
