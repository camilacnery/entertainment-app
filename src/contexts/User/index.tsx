import { ReactNode, FC, createContext, useState, useContext, useEffect } from "react";
import { IUserProfile } from "@/domain/User";
import { decodeJwt } from "@/utils/jwt";

interface IDecodedUser {
  aud: string;
  azp: string;
  email: string;
  email_verified: boolean;
  exp: number;
  family_name: string;
  given_name: string;
  iat: number;
  iss: string;
  jti: string;
  name: string;
  nbf: number;
  picture: string;
  sub: string;
}

export type TUserContext = {
  profile?: IUserProfile;
  onSuccessLogin: (userJwt: string) => void;
  onLogout: () => void;
};

const UserContext = createContext<TUserContext>({} as TUserContext);

type Props = {
  children: ReactNode;
};

const STORAGE_ID = "app/user-token";

const UserContextProvider: FC<Props> = ({ children }) => {
  const [profile, setProfile] = useState<IUserProfile>();

  const setProfileFromToken = (userJwt: string) => {
    const profileData = decodeJwt<IDecodedUser>(userJwt);

    setProfile({
      fullName: profileData.name,
      firstName: profileData.given_name,
      email: profileData.email,
      imageUrl: profileData.picture,
    });
  };

  const onSuccessLogin = (userJwt: string) => {
    localStorage.setItem(STORAGE_ID, userJwt);
    setProfileFromToken(userJwt);
  };

  const onLogout = () => {
    localStorage.removeItem(STORAGE_ID);
    setProfile(undefined);
  };

  useEffect(() => {
    const userJwt = localStorage.getItem(STORAGE_ID);
    if (!userJwt) return;

    setProfileFromToken(userJwt);
  }, []);

  return (
    <UserContext.Provider
      value={{
        profile,
        onSuccessLogin,
        onLogout,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

const useUserContext = () => useContext(UserContext);

export { useUserContext, UserContext, UserContextProvider };
