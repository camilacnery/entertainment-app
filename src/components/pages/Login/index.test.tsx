import { UserContextProvider } from "@/contexts/User";
import { render, screen } from "@testing-library/react";
import { CredentialResponse, GoogleLoginProps } from "@react-oauth/google";
import userEvent from "@testing-library/user-event";
import LoginPage from ".";

const mockedRouterPush = jest.fn();
jest.mock("next/router", () => ({
  useRouter: () => ({
    push: mockedRouterPush,
  }),
}));

jest.mock("@react-oauth/google", () => {
  const GoogleAuth = jest.requireActual("@react-oauth/google");

  const credentialsResponse: CredentialResponse = {
    credential:
      "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJFbnRlcnRhaW5tZW50IEFwcCIsImlhdCI6MTY2NDc0MzE3MCwiZXhwIjoxNjk2Mjc5MTcwLCJhdWQiOiJ3d3cuZXhhbXBsZS5jb20iLCJzdWIiOiJ0ZXN0QGV4YW1wbGUuY29tIiwiZW1haWwiOiJ0ZXN0QGV4YW1wbGUuY29tIiwiYXpwIjoiaGphc29pa2pscWpubGsiLCJlbWFpbF92ZXJpZmllZCI6InRydWUiLCJmYW1pbHlfbmFtZSI6IlRlc3QiLCJnaXZlbl9uYW1lIjoiVXNlciIsImp0aSI6Impsa2phbHNqZGxhamRsYWoiLCJuYW1lIjoiVGVzdCBVc2VyIiwibmJmIjoiamxja3NhamxkamFzbGQiLCJwaWN0dXJlIjoiaHR0cHM6Ly90ZXN0LnBpY3R1cmUifQ.DHRCtB-XkEcNaULDVjLq_ZHV28ocAYVtHmQLKYBAhmI",
  };

  const GoogleLogin = ({ onSuccess }: GoogleLoginProps) => (
    <button onClick={() => onSuccess(credentialsResponse)}>Mock login</button>
  );
  return { ...GoogleAuth, GoogleLogin };
});

const pageElements = {
  loggedInMessage: () => screen.getByRole("heading", { name: "You are logged in, redirecting..." }),
  mockLoginButton: () => screen.getByRole("button", { name: "Mock login" }),
};

const renderLogin = () => {
  return render(
    <UserContextProvider>
      <LoginPage />
    </UserContextProvider>
  );
};

describe("Login Page", () => {
  beforeEach(() => {
    window.localStorage.clear();
  });

  it("renders login page", async () => {
    const { baseElement } = renderLogin();

    expect(baseElement).toMatchSnapshot();

    await userEvent.click(pageElements.mockLoginButton());
    expect(pageElements.loggedInMessage()).toBeInTheDocument();
    expect(mockedRouterPush).toHaveBeenCalledWith("/");
  });
});
