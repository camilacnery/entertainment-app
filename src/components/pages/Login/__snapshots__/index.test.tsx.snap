// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page renders login page 1`] = `
<body>
  <div>
    <div
      class="banner"
    />
    <main
      class="main"
    >
      <div
        class="content"
      >
        <section
          class="content"
        >
          <h1>
            Login with Google
          </h1>
          <button>
            Mock login
          </button>
        </section>
      </div>
    </main>
  </div>
</body>
`;
